torch/csrc/autograd/python_function.cpp

```diff
@@ -635,7 +637,10 @@ PyObject *THPFunction_do_forward(THPFunction *self, PyObject *_inputs)
   auto& unpacked_input = info_pair.first;
   auto& input_info = info_pair.second;
   bool is_executable = input_info.is_executable;
-  self->cdata.set_next_edges(std::move(input_info.next_edges));
+  Py_INCREF(self);
+  auto cdata = std::shared_ptr<PyFunction>(new PyFunction(THPObjectPtr((PyObject*)self)), deleteFunction);
+  self->cdata = cdata;
```

Comment:
Is it possible that this method is called multiple times on the same `THPFunction`, so that in some invocations `self->cdata` is already non-NULL?

Reply:
Yeah, in the legacy case it might be non-NULL. I actually have no idea what the "right" thing to do here is, because legacy autograd function behavior is nuts (all the better for @yf225 to delete them). Check this out:

```
macbook-pro-116:~ ezyang$ cat abb.py
import torch
from torch.autograd import Function
class Id(Function):
    def forward(self, x):
        self.save_for_backward(x)
        return x

    def backward(self, grad_x):
        x = self.saved_tensors
        print(x)
        return x

f = Id()
x1 = torch.zeros(1, requires_grad=True)
x2 = torch.ones(1, requires_grad=True)
print(f)
y = f(x1)
print(y.grad_fn)
z = f(x2)
print(z.grad_fn)
y.backward()

macbook-pro-116:~ ezyang$ python abb.py
<__main__.Id object at 0x116d58850>
<__main__.Id object at 0x116d58850>
<__main__.Id object at 0x116d58850>
(tensor([1.], requires_grad=True),)
```

The second call to f clobbers the saved tensors of the first call (in y), because f is *literally* the same `grad_fn` in both cases, even though this is totally wrong. I think the API is misused here, for a legacy Function, you should only call `__apply__` on it once.

On reflection, maybe the right thing is to not allocate a new `PyFunction` if there already is one. I can make this adjustment, but really this program is stupid and we shouldn't support it.

Reply:
I've modified legacy to warn when this happens.